File: backend/src/main/java/com/ia/app/util/CpfCnpjValidator.java
package com.ia.app.util;

public class CpfCnpjValidator {
  private CpfCnpjValidator() {}

  public static boolean isValid(String value) {
    if (value == null) return false;
    String digits = value.replaceAll("\\D", "");
    if (digits.length() == 11) return isValidCpf(digits);
    if (digits.length() == 14) return isValidCnpj(digits);
    return false;
  }

  private static boolean isValidCpf(String cpf) {
    if (cpf.matches("(\\d)\\1{10}")) return false;
    int sum = 0;
    for (int i = 0; i < 9; i++) sum += (cpf.charAt(i) - '0') * (10 - i);
    int dv = (sum * 10) % 11;
    if (dv == 10) dv = 0;
    if (dv != (cpf.charAt(9) - '0')) return false;
    sum = 0;
    for (int i = 0; i < 10; i++) sum += (cpf.charAt(i) - '0') * (11 - i);
    dv = (sum * 10) % 11;
    if (dv == 10) dv = 0;
    return dv == (cpf.charAt(10) - '0');
  }

  private static boolean isValidCnpj(String cnpj) {
    if (cnpj.matches("(\\d)\\1{13}")) return false;
    int[] w1 = {5,4,3,2,9,8,7,6,5,4,3,2};
    int[] w2 = {6,5,4,3,2,9,8,7,6,5,4,3,2};
    int sum = 0;
    for (int i = 0; i < 12; i++) sum += (cnpj.charAt(i) - '0') * w1[i];
    int dv = sum % 11;
    dv = dv < 2 ? 0 : 11 - dv;
    if (dv != (cnpj.charAt(12) - '0')) return false;
    sum = 0;
    for (int i = 0; i < 13; i++) sum += (cnpj.charAt(i) - '0') * w2[i];
    dv = sum % 11;
    dv = dv < 2 ? 0 : 11 - dv;
    return dv == (cnpj.charAt(13) - '0');
  }
}

File: frontend/src/app/shared/cpf-cnpj.validator.ts
import { AbstractControl, ValidationErrors } from '@angular/forms';

function onlyDigits(value: string): string {
  return value.replace(/\D/g, '');
}

function isValidCpf(cpf: string): boolean {
  if (cpf.length !== 11 || /^([0-9])\1+$/.test(cpf)) return false;
  let sum = 0;
  for (let i = 0; i < 9; i++) sum += parseInt(cpf.charAt(i), 10) * (10 - i);
  let dv = (sum * 10) % 11;
  if (dv === 10) dv = 0;
  if (dv !== parseInt(cpf.charAt(9), 10)) return false;
  sum = 0;
  for (let i = 0; i < 10; i++) sum += parseInt(cpf.charAt(i), 10) * (11 - i);
  dv = (sum * 10) % 11;
  if (dv === 10) dv = 0;
  return dv === parseInt(cpf.charAt(10), 10);
}

function isValidCnpj(cnpj: string): boolean {
  if (cnpj.length !== 14 || /^([0-9])\1+$/.test(cnpj)) return false;
  const weights1 = [5,4,3,2,9,8,7,6,5,4,3,2];
  const weights2 = [6,5,4,3,2,9,8,7,6,5,4,3,2];
  let sum = 0;
  for (let i = 0; i < 12; i++) sum += parseInt(cnpj.charAt(i), 10) * (weights1[i] ?? 0);
  let dv = sum % 11;
  dv = dv < 2 ? 0 : 11 - dv;
  if (dv !== parseInt(cnpj.charAt(12), 10)) return false;
  sum = 0;
  for (let i = 0; i < 13; i++) sum += parseInt(cnpj.charAt(i), 10) * (weights2[i] ?? 0);
  dv = sum % 11;
  dv = dv < 2 ? 0 : 11 - dv;
  return dv === parseInt(cnpj.charAt(13), 10);
}

export function cpfCnpjValidator(control: AbstractControl): ValidationErrors | null {
  const value = (control.value || '').toString().trim();
  if (!value) return null;
  const digits = onlyDigits(value);
  const valid = digits.length === 11 ? isValidCpf(digits) : isValidCnpj(digits);
  return valid ? null : { cpfCnpj: true };
}

export function formatCpfCnpj(value: string): string {
  const digits = onlyDigits(value);
  if (digits.length <= 11) {
    return digits
      .replace(/(\d{3})(\d)/, '$1.$2')
      .replace(/(\d{3})(\d)/, '$1.$2')
      .replace(/(\d{3})(\d{1,2})$/, '$1-$2');
  }
  return digits
    .replace(/(\d{2})(\d)/, '$1.$2')
    .replace(/(\d{3})(\d)/, '$1.$2')
    .replace(/(\d{3})(\d)/, '$1/$2')
    .replace(/(\d{4})(\d{1,2})$/, '$1-$2');
}
